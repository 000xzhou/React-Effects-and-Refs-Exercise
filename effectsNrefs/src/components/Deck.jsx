import { useState, useEffect } from "react";
import axios from "axios";
import Card from "./Card";

function Deck() {
  const [card, setCard] = useState([]);
  const [deck, setDeck] = useState();

  const DECK_URL = "https://deckofcardsapi.com/api/deck";

  // shuffle deck
  const shuffleDeck = async () => {
    await axios.get(`${DECK_URL}/${deck}/shuffle/`);
    setCard([]);
  };
  // get new card from deck
  const getNewCard = async () => {
    const cardType = await axios.get(`${DECK_URL}/${deck}/draw/?count=1`);
    setCard((card) => [
      ...card,
      {
        id: cardType.data.cards[0].code,
        name: cardType.data.cards[0].suit + " " + cardType.data.cards[0].value,
        image: cardType.data.cards[0].image,
      },
    ]);
  };

  //   get new deck once at the start of game
  useEffect(() => {
    const fetchDeck = async () => {
      const deckId = await axios.get(`${DECK_URL}/new/shuffle/?deck_count=1`);
      setDeck(deckId.data.deck_id);
    };
    fetchDeck();
  }, []);

  return (
    <>
      <button onClick={shuffleDeck}>New Deck</button>
      <button onClick={getNewCard}>Draw Card</button>
      <div style={{ display: "flex", flexWrap: "wrap" }}>
        {card.map((c) => (
          <Card key={c.id} image={c.image} name={c.name} />
        ))}
      </div>
    </>
  );
}

export default Deck;
