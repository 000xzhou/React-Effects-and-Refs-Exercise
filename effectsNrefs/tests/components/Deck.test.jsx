import { describe, it, expect, vi } from "vitest";
import { render, fireEvent, waitFor, act } from "@testing-library/react";
import Deck from "../../src/components/Deck";
import axios from "axios";

// Mock axios
vi.mock("axios");

const DECK_URL = "https://deckofcardsapi.com/api/deck";

// Mock data for a new deck and drawing a card
const mockDeckResponse = {
  data: {
    deck_id: "testdeckid",
    shuffled: true,
    remaining: 52,
  },
};

const mockCardResponse = {
  data: {
    success: true,
    deck_id: "testdeckid",
    cards: [
      {
        code: "6H",
        image: "https://deckofcardsapi.com/static/img/6H.png",
        images: {
          svg: "https://deckofcardsapi.com/static/img/6H.svg",
          png: "https://deckofcardsapi.com/static/img/6H.png",
        },
        value: "6",
        suit: "HEARTS",
      },
    ],
    remaining: 50,
  },
};

// Mock the axios.get method
axios.get.mockImplementation((url) => {
  if (url === `${DECK_URL}/new/shuffle/?deck_count=1`) {
    return Promise.resolve(mockDeckResponse);
  } else if (url === `${DECK_URL}/testdeckid/draw/?count=1`) {
    return Promise.resolve(mockCardResponse);
  }
  return Promise.reject(new Error("not found"));
});

describe("Deck", () => {
  it("renders the Deck component", async () => {
    await act(async () => {
      render(<Deck />);
    });
  });

  it("should match snapshot", async () => {
    let asFragment;
    await act(async () => {
      const rendered = render(<Deck />);
      asFragment = rendered.asFragment;
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("should draw a card", async () => {
    let getByText, getByAltText;
    await act(async () => {
      const rendered = render(<Deck />);
      getByText = rendered.getByText;
      getByAltText = rendered.getByAltText;
    });

    // draw card
    const drawCardButton = getByText("Draw Card");
    await act(async () => {
      fireEvent.click(drawCardButton);
    });
    // waitFor to wait for the asynchronous update of the DOM.
    await waitFor(() => {
      const cardImage = getByAltText("HEARTS 6");
      expect(cardImage.src).toBe(
        "https://deckofcardsapi.com/static/img/6H.png"
      );
    });
  });
});
