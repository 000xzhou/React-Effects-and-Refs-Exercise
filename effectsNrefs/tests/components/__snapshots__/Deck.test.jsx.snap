// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Deck > should match snapshot 1`] = `
<DocumentFragment>
  <button>
    New Deck
  </button>
  <button>
    Draw Card
  </button>
  <div
    style="display: flex; flex-wrap: wrap;"
  />
</DocumentFragment>
`;
